### PR TITLE
Support wildcard/regex for indices param in _remotestore/_restore

### PR DESCRIPTION
**Pull request by linuxpi**
### Description
- _remotestore/_restore will now support wildcards/regex for indices param.
- Specifying body - `{ "indices": ["my-index-*"] }` will match all indices whose name start with `my-index-` and trigger remote store restore.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/8480

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by BhumikaSaini-Amazon — CHANGES_REQUESTED** [review]
Thanks @linuxpi!
Please could you also enable the ITs in `RemoteStoreIT` that are currently awaiting fix on #8480? We need to pass those tests as part of this enhancement.

**Comment by linuxpi** [comment]
### Test Failures

[org.opensearch.index.codec.MultiCodecReindexIT.testReindexingMultipleCodecs](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.index.codec/MultiCodecReindexIT/testReindexingMultipleCodecs/)
    [org.opensearch.index.codec.MultiCodecReindexIT.classMethod](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.index.codec/MultiCodecReindexIT/classMethod/)
    [org.opensearch.backwards.IndexingIT.testIndexVersionPropagation](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/IndexingIT/testIndexVersionPropagation/)
    [org.opensearch.backwards.IndexingIT.testSeqNoCheckpoints](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/IndexingIT/testSeqNoCheckpoints/)
    [org.opensearch.backwards.IndexingIT.testIndexingWithReplicaOnBwcNodes](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/IndexingIT/testIndexingWithReplicaOnBwcNodes/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/20_translog/Translog stats on closed indices}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_20_translog_Translog_stats_on_closed_indices_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Fields - pattern}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Fields___pattern_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Completion - pattern}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Completion___pattern_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Completion fields - multi}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Completion_fields___multi_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=cat.nodes/10_basic/Additional disk information}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_cat_nodes_10_basic_Additional_disk_information_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Fields - star}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Fields___star_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Completion fields - star}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Completion_fields___star_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/10_basic/Nodes stats level}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_nodes_stats_10_basic_Nodes_stats_level_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Fielddata fields - pattern}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Fielddata_fields___pattern_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=nodes.stats/11_indices_metrics/Metric - indices _all}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_nodes_stats_11_indices_metrics_Metric___indices__all_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Fields - multi}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Fields___multi_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Completion - all metric}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Completion___all_metric_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/14_groups/Groups - blank}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_14_groups_Groups___blank_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/12_level/Level - blank}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_12_level_Level___blank_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=indices.stats/13_fields/Fielddata fields - one}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_indices_stats_13_fields_Fielddata_fields___one_/)
    [org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=cat.indices/20_hidden/Test cat indices output with a hidden index with a hidden alias}](https://build.ci.opensearch.org/job/gradle-check/25013/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_cat_indices_20_hidden_Test_cat_indices_output_with_a_hidden_index_with_a_hidden_alias_/)

**Comment by linuxpi** [comment]
[Test Result](https://build.ci.opensearch.org/job/gradle-check/25040/testReport/) (1 failure / +1)

    [org.opensearch.search.profile.query.QueryProfilerIT.testProfileMatchesRegular {p0={"search.concurrent_segment_search.enabled":"true"} p1=true}](https://build.ci.opensearch.org/job/gradle-check/25040/testReport/junit/org.opensearch.search.profile.query/QueryProfilerIT/testProfileMatchesRegular__p0___search_concurrent_segment_search_enabled___true___p1_true_/)

this is flaky https://github.com/opensearch-project/OpenSearch/issues/9869

Did Local retries on source branch. 2 failures in 110 tests. But different tests than what failed here

```
./gradlew ':server:internalClusterTest' --tests "org.opensearch.search.profile.query.QueryProfilerIT" -Dtests.iters=5

Tests with failures:
 - org.opensearch.search.profile.query.QueryProfilerIT.testProfileQuery {p0={"search.concurrent_segment_search.enabled":"true"} p1=true seed=[F3EFCB81911DEB22:39EA9A865650C0D]}
 - org.opensearch.search.profile.query.QueryProfilerIT.testProfileQuery {p0={"search.concurrent_segment_search.enabled":"true"} p1=true seed=[F3EFCB81911DEB22:4F5F2C05C46DCCB6]}

110 tests completed, 2 failed

```

**Comment by linuxpi** [comment]
```
[Test Result](https://build.ci.opensearch.org/job/gradle-check/32484/testReport/) (3 failures / +2)

    [org.opensearch.http.SearchRestCancellationIT.testAutomaticCancellationMultiSearchDuringFetchPhase](https://build.ci.opensearch.org/job/gradle-check/32484/testReport/junit/org.opensearch.http/SearchRestCancellationIT/testAutomaticCancellationMultiSearchDuringFetchPhase/)
    [org.opensearch.cluster.MinimumClusterManagerNodesIT.testThreeNodesNoClusterManagerBlock](https://build.ci.opensearch.org/job/gradle-check/32484/testReport/junit/org.opensearch.cluster/MinimumClusterManagerNodesIT/testThreeNodesNoClusterManagerBlock/)
    [org.opensearch.remotestore.RemoteIndexPrimaryRelocationIT.testPrimaryRelocationWhileIndexing](https://build.ci.opensearch.org/job/gradle-check/32484/testReport/junit/org.opensearch.remotestore/RemoteIndexPrimaryRelocationIT/testPrimaryRelocationWhileIndexing/)
```

**Comment by linuxpi** [comment]
```
[Test Result](https://build.ci.opensearch.org/job/gradle-check/32486/testReport/) (1 failure / +1)

    [org.opensearch.cluster.decommission.DecommissionControllerTests.testTimesOut](https://build.ci.opensearch.org/job/gradle-check/32486/testReport/junit/org.opensearch.cluster.decommission/DecommissionControllerTests/testTimesOut/)
```

**Comment by peternied** [comment]
@linuxpi Thanks for the updates, could you address the failing checks?  @BhumikaSaini-Amazon Could you take another look since the updates have come in?

File: CHANGELOG.md
Comment (BhumikaSaini-Amazon) [inline CHANGELOG.md:220]:
```suggestion
- Support wildcard/regex for indices param in `_remotestore/_restore` ([#8922](https://github.com/opensearch-project/OpenSearch/pull/8922))
```

**Review by BhumikaSaini-Amazon — APPROVED** [review]

**Comment by peternied** [comment]
@linuxpi looks like there are some merge conflicts that need to be resolved, can you take a look?

**Comment by linuxpi** [comment]
> @linuxpi looks like there are some merge conflicts that need to be resolved, can you take a look?

@peternied these are in CHANGELOG and keeps coming up again and again as i rebase, i am waiting on getting a maintainer approval before rebasing as there is a good change i would have rebase again if i do it now

**Comment by peternied** [comment]
@linuxpi If you are waiting on approval then you are going to be stuck for a while.  When there are conflicts, the CI checks cannot be run.  Until there are green CI checks - I will not approve a change, this is often the case with other maintainers.

If you feel that the changelog is constantly having merge conflicts consider moving your changelog entry to the top of the list where a conflict is less likely.

**Comment by dblock** [comment]
@andrross or @mch2 care to CR?

File: server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java:53]: nit: Move it to common utils

**Comment by ashking94** [comment]
[Storage Triage - attendees [1](https://github.com/ashking94/) [2](https://github.com/sachinpkale/) [3](https://github.com/gbbafna/) [4](https://github.com/Bukhtawar/) [5](https://github.com/rramachand21/) [6](https://github.com/skumawat2025/) [7](https://github.com/shourya035/) [8](https://github.com/vikasvb90/) [9](https://github.com/BhumikaSaini-Amazon/) [10](https://github.com/ltaragi/) [11](https://github.com/lukas-vlcek/) [12](https://github.com/astute-decipher/)]

@linuxpi  Thanks for your contribution. Please feel free to comeback and bring this to closure. If you have any questions, do take it up with the reviewers/maintainers.

**Review by peternied — COMMENTED** [review]
@linuxpi I've cleaned up the conflict, and verify this is in a green state, before we merge this improvement please add a test case(s) that verifies the regex/wildcard scenario.

**Review by sachinpkale — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java
Comment (linuxpi) [reply server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java]: Added IndexUtils in common for this

**Comment by linuxpi** [comment]
> @linuxpi I've cleaned up the conflict, and verify this is in a green state, before we merge this improvement please add a test case(s) that verifies the regex/wildcard scenario.

Integ tests for this already existed before my change but were skipped due to AwaitsFix annotation. I've remove this and now these tests are run as part of gradle check.

**Comment by linuxpi** [comment]
Flaky Test - https://github.com/opensearch-project/OpenSearch/issues/10558


[org.opensearch.cluster.allocation.ClusterRerouteIT.testDelayWithALargeAmountOfShards](https://build.ci.opensearch.org/job/gradle-check/38198/testReport/junit/org.opensearch.cluster.allocation/ClusterRerouteIT/testDelayWithALargeAmountOfShards/)

**Review by peternied — APPROVED** [review]

**Comment by peternied** [comment]
@linuxpi Please investigate the tests failures after the results come back:

```
* What went wrong:
Execution failed for task ':distribution:bwc:minor:buildBwcLinuxTar'.
> Building 2.14.0 didn't generate expected file /var/jenkins/workspace/gradle-check/search/distribution/bwc/minor/build/bwc/checkout-2.x/distribution/archives/linux-tar/build/distributions/opensearch-min-2.14.0-SNAPSHOT-linux-x64.tar.gz
```

Check will keep failing until you've rebased.

**Review by sachinpkale — APPROVED** [review]

**Comment by linuxpi** [comment]
backport pr https://github.com/opensearch-project/OpenSearch/pull/13527

**Pull request merged**
